Đọc dữ liệu từ file, tìm hiểu dữ liệu 

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('hcvdat0.csv')

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 615 entries, 0 to 614
Data columns (total 14 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  615 non-null    int64  
 1   Category    615 non-null    object 
 2   Age         615 non-null    int64  
 3   Sex         615 non-null    object 
 4   ALB         614 non-null    float64
 5   ALP         597 non-null    float64
 6   ALT         614 non-null    float64
 7   AST         615 non-null    float64
 8   BIL         615 non-null    float64
 9   CHE         615 non-null    float64
 10  CHOL        605 non-null    float64
 11  CREA        615 non-null    float64
 12  GGT         615 non-null    float64
 13  PROT        614 non-null    float64
dtypes: float64(10), int64(2), object(2)
memory usage: 67.4+ KB


<b>Tập dữ liệu HCV gồm 615 phần tử (hàng) và 13 thuộc tính (cột)</b>
<br />
<b>Tập dữ liệu có lẽ được lấy từ các bệnh nhân hiến máu tự nguyện, sau khi hiến máu thì bác sĩ thường sẽ tiến hành xét nghiệm để loại ra máu của người bệnh</b>

<b>Trong đó có cột Category là nhãn của tập dữ liệu này (đầu ra) => Bài toán phân lớp => Xây dựng mô hình phân lớp để dự đoán</b>
<b>Các cột còn lại là đầu vào</b>
<b>Ngoài ra tập dữ liệu còn có thể được sử dụng để gom nhóm (Clustering) các bệnh nhân, áp dụng trích chọn đặc trưng (Feature Selection) để chọn lọc các chỉ số phù hợp nhất để dự đoán bệnh, giảm chiều cho tập dữ liệu</b>

In [4]:
df.head()

Unnamed: 0       Category  Age Sex   ALB   ALP   ALT   AST   BIL    CHE  \
0           1  0=Blood Donor   32   m  38.5  52.5   7.7  22.1   7.5   6.93   
1           2  0=Blood Donor   32   m  38.5  70.3  18.0  24.7   3.9  11.17   
2           3  0=Blood Donor   32   m  46.9  74.7  36.2  52.6   6.1   8.84   
3           4  0=Blood Donor   32   m  43.2  52.0  30.6  22.6  18.9   7.33   
4           5  0=Blood Donor   32   m  39.2  74.1  32.6  24.8   9.6   9.15   

   CHOL   CREA   GGT  PROT  
0  3.23  106.0  12.1  69.0  
1  4.80   74.0  15.6  76.5  
2  5.20   86.0  33.2  79.3  
3  4.74   80.0  33.8  75.7  
4  4.32   76.0  29.9  68.7

In [5]:
# unique() hien thi cac gia tri duy nhat khong lap lai cua mot thuoc tinh
# nunique() dem so cac gia tri duy nhat khong lap lai cua mot thuoc tinh
# np.sort() sap xep mang theo gia tri tang dan

df['Category'].unique()

array(['0=Blood Donor', '0s=suspect Blood Donor', '1=Hepatitis',
       '2=Fibrosis', '3=Cirrhosis'], dtype=object)

In [7]:
# Count the number of missing values in each column
missing_values = df.isnull().sum()

# Print the number of missing values in each column
print("Số lượng dữ liệu bị thiếu ở mỗi cột:")
print(missing_values)

Số lượng dữ liệu bị thiếu ở mỗi cột:
Unnamed: 0     0
Category       0
Age            0
Sex            0
ALB            1
ALP           18
ALT            1
AST            0
BIL            0
CHE            0
CHOL          10
CREA           0
GGT            0
PROT           1
dtype: int64


Tập dữ liệu có 5 nhãn: Blood Donor là nhóm máu hiến bình thường (không bị bệnh về gan, có thể bị bệnh khác), suspect Blood Donor có lẽ là nhóm mất nhãn không xác định được, Hepatitis là nhóm người bệnh viêm gan C, Fibrosis là nhóm người bệnh xơ gan, Cirrhosis là nhóm người bị chai gan

Bỏ cột 0s=suspect Blood Donor do có quá ít dữ liệu, dữ liệu không liên quan và có nhiều dữ liệu nhiễu

In [42]:
df_filtered = df[df['Category'] != '0s=suspect Blood Donor']

In [43]:
df_filtered['Category'].unique()

array(['0=Blood Donor', '1=Hepatitis', '2=Fibrosis', '3=Cirrhosis'],
      dtype=object)

Sử dụng one hot encoding để chuyển cột Sex về giá trị Boolean

In [44]:
# Perform one-hot encoding
one_hot_encoded = pd.get_dummies(df_filtered['Sex'])

# Concatenate the one-hot encoded columns with the original DataFrame
df_filtered = pd.concat([df_filtered, one_hot_encoded], axis=1)

# Remove one of the encoded columns (e.g., remove 'm' column)
df_filtered.drop('f', axis=1, inplace=True)

# Optionally, you can also drop the original 'Sex' column
df_filtered.drop('Sex', axis=1, inplace=True)

# Display the resulting DataFrame
print(df_filtered)

     Unnamed: 0       Category  Age   ALB    ALP    ALT    AST   BIL    CHE  \
0             1  0=Blood Donor   32  38.5   52.5    7.7   22.1   7.5   6.93   
1             2  0=Blood Donor   32  38.5   70.3   18.0   24.7   3.9  11.17   
2             3  0=Blood Donor   32  46.9   74.7   36.2   52.6   6.1   8.84   
3             4  0=Blood Donor   32  43.2   52.0   30.6   22.6  18.9   7.33   
4             5  0=Blood Donor   32  39.2   74.1   32.6   24.8   9.6   9.15   
..          ...            ...  ...   ...    ...    ...    ...   ...    ...   
610         611    3=Cirrhosis   62  32.0  416.6    5.9  110.3  50.0   5.57   
611         612    3=Cirrhosis   64  24.0  102.8    2.9   44.4  20.0   1.54   
612         613    3=Cirrhosis   64  29.0   87.3    3.5   99.0  48.0   1.66   
613         614    3=Cirrhosis   46  33.0    NaN   39.0   62.0  20.0   3.56   
614         615    3=Cirrhosis   59  36.0    NaN  100.0   80.0  12.0   9.07   

     CHOL   CREA    GGT  PROT      m  
0    3.23  1

Đổi tên cột Sex mới (m) là nam hay không thành isMale

In [45]:
# Rename the remaining column to 'isFemale'
df_filtered.rename(columns={'m': 'isMale'}, inplace=True)

# Display the resulting DataFrame
print(df_filtered)

     Unnamed: 0       Category  Age   ALB    ALP    ALT    AST   BIL    CHE  \
0             1  0=Blood Donor   32  38.5   52.5    7.7   22.1   7.5   6.93   
1             2  0=Blood Donor   32  38.5   70.3   18.0   24.7   3.9  11.17   
2             3  0=Blood Donor   32  46.9   74.7   36.2   52.6   6.1   8.84   
3             4  0=Blood Donor   32  43.2   52.0   30.6   22.6  18.9   7.33   
4             5  0=Blood Donor   32  39.2   74.1   32.6   24.8   9.6   9.15   
..          ...            ...  ...   ...    ...    ...    ...   ...    ...   
610         611    3=Cirrhosis   62  32.0  416.6    5.9  110.3  50.0   5.57   
611         612    3=Cirrhosis   64  24.0  102.8    2.9   44.4  20.0   1.54   
612         613    3=Cirrhosis   64  29.0   87.3    3.5   99.0  48.0   1.66   
613         614    3=Cirrhosis   46  33.0    NaN   39.0   62.0  20.0   3.56   
614         615    3=Cirrhosis   59  36.0    NaN  100.0   80.0  12.0   9.07   

     CHOL   CREA    GGT  PROT  isMale  
0    3.23  

Bỏ phần tử thiếu

In [46]:
# Drop all rows containing NA values
df_filtered = df_filtered.dropna()

# Display the resulting DataFrame
print(df_filtered)

     Unnamed: 0       Category  Age   ALB    ALP   ALT    AST   BIL    CHE  \
0             1  0=Blood Donor   32  38.5   52.5   7.7   22.1   7.5   6.93   
1             2  0=Blood Donor   32  38.5   70.3  18.0   24.7   3.9  11.17   
2             3  0=Blood Donor   32  46.9   74.7  36.2   52.6   6.1   8.84   
3             4  0=Blood Donor   32  43.2   52.0  30.6   22.6  18.9   7.33   
4             5  0=Blood Donor   32  39.2   74.1  32.6   24.8   9.6   9.15   
..          ...            ...  ...   ...    ...   ...    ...   ...    ...   
608         609    3=Cirrhosis   58  34.0   46.4  15.0  150.0   8.0   6.26   
609         610    3=Cirrhosis   59  39.0   51.3  19.6  285.8  40.0   5.77   
610         611    3=Cirrhosis   62  32.0  416.6   5.9  110.3  50.0   5.57   
611         612    3=Cirrhosis   64  24.0  102.8   2.9   44.4  20.0   1.54   
612         613    3=Cirrhosis   64  29.0   87.3   3.5   99.0  48.0   1.66   

     CHOL   CREA    GGT  PROT  isMale  
0    3.23  106.0   12.1

Bỏ cột unnamed (Cột đầu) do không dùng để xây dựng mô hình

In [47]:
df_filtered = df_filtered.drop(df.columns[0], axis=1)

In [48]:
df_filtered.head()

Category  Age   ALB   ALP   ALT   AST   BIL    CHE  CHOL   CREA   GGT  \
0  0=Blood Donor   32  38.5  52.5   7.7  22.1   7.5   6.93  3.23  106.0  12.1   
1  0=Blood Donor   32  38.5  70.3  18.0  24.7   3.9  11.17  4.80   74.0  15.6   
2  0=Blood Donor   32  46.9  74.7  36.2  52.6   6.1   8.84  5.20   86.0  33.2   
3  0=Blood Donor   32  43.2  52.0  30.6  22.6  18.9   7.33  4.74   80.0  33.8   
4  0=Blood Donor   32  39.2  74.1  32.6  24.8   9.6   9.15  4.32   76.0  29.9   

   PROT  isMale  
0  69.0    True  
1  76.5    True  
2  79.3    True  
3  75.7    True  
4  68.7    True

Đổi tên cột nhãn cho dễ đọc

In [49]:
df_filtered['Category'] = df_filtered['Category'].replace('0=Blood Donor', 'Blood Donor')
df_filtered['Category'] = df_filtered['Category'].replace('1=Hepatitis', 'Hepatitis')
df_filtered['Category'] = df_filtered['Category'].replace('2=Fibrosis', 'Fibrosis')
df_filtered['Category'] = df_filtered['Category'].replace('3=Cirrhosis', 'Cirrhosis')

In [50]:
df_filtered.head()

Category  Age   ALB   ALP   ALT   AST   BIL    CHE  CHOL   CREA   GGT  \
0  Blood Donor   32  38.5  52.5   7.7  22.1   7.5   6.93  3.23  106.0  12.1   
1  Blood Donor   32  38.5  70.3  18.0  24.7   3.9  11.17  4.80   74.0  15.6   
2  Blood Donor   32  46.9  74.7  36.2  52.6   6.1   8.84  5.20   86.0  33.2   
3  Blood Donor   32  43.2  52.0  30.6  22.6  18.9   7.33  4.74   80.0  33.8   
4  Blood Donor   32  39.2  74.1  32.6  24.8   9.6   9.15  4.32   76.0  29.9   

   PROT  isMale  
0  69.0    True  
1  76.5    True  
2  79.3    True  
3  75.7    True  
4  68.7    True

In [51]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import f1_score, precision_score, recall_score
from sklearn.metrics import confusion_matrix
import numpy as np

<h2>SVM</h2>

In [53]:
# Separate features (X) and target variable (y)
X = df_filtered.drop(columns=['Category'])
y = df_filtered['Category']

# Loop to train the model ten times
f1_scores = []
precision_scores = []
recall_scores = []

for i in range(20):
    # Split data into train and test sets
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=i)
    
    # Scale features
    scaler = StandardScaler()
    X_train_scaled = scaler.fit_transform(X_train)
    X_test_scaled = scaler.transform(X_test)
    
    # Train decision tree classifier
    classifier = DecisionTreeClassifier()
    classifier.fit(X_train_scaled, y_train)
    
    # Predict on the test set
    y_pred = classifier.predict(X_test_scaled)
    
    # Calculate evaluation metrics
    f1 = f1_score(y_test, y_pred, average='weighted')
    precision = precision_score(y_test, y_pred, average='weighted')
    recall = recall_score(y_test, y_pred, average='weighted')
    
    # Append scores to lists
    f1_scores.append(f1)
    precision_scores.append(precision)
    recall_scores.append(recall)

    # Confusion matrix
    cm = confusion_matrix(y_test, y_pred)
    
    # Print scores for each loop
    print(f"Loop {i + 1}: F1 Score: {f1}, Precision: {precision}, Recall: {recall}")
    print("Confusion Matrix:")
    print(cm)
    print()

# Calculate average scores
avg_f1 = np.mean(f1_scores)
avg_precision = np.mean(precision_scores)
avg_recall = np.mean(recall_scores)

# Print average scores
print(f"Average F1 Score: {avg_f1}")
print(f"Average Precision: {avg_precision}")
print(f"Average Recall: {avg_recall}")

Loop 1: F1 Score: 0.9502053502053502, Precision: 0.9476159759178627, Recall: 0.9572649572649573
Confusion Matrix:
[[104   0   0   0]
 [  0   5   0   0]
 [  2   0   1   1]
 [  0   1   1   2]]

Loop 2: F1 Score: 0.9701646090534979, Precision: 0.9746726260487729, Recall: 0.9743589743589743
Confusion Matrix:
[[107   0   0   0]
 [  0   2   0   0]
 [  1   0   1   1]
 [  1   0   0   4]]

Loop 3: F1 Score: 0.9701048072088797, Precision: 0.9658889658889659, Recall: 0.9743589743589743
Confusion Matrix:
[[110   0   0   0]
 [  0   2   0   0]
 [  1   0   0   1]
 [  0   0   1   2]]

Loop 4: F1 Score: 0.9840455840455841, Precision: 0.9871794871794872, Recall: 0.9829059829059829
Confusion Matrix:
[[107   0   0   0]
 [  0   7   1   0]
 [  0   0   0   1]
 [  0   0   0   1]]

Loop 5: F1 Score: 0.9451333579773947, Precision: 0.934071484071484, Recall: 0.9572649572649573
Confusion Matrix:
[[108   0   0   0]
 [  0   3   0   0]
 [  1   0   0   2]
 [  1   1   0   1]]

Loop 6: F1 Score: 0.9517441936796776, Pre

c:\users\hp\appdata\local\programs\python\python39\lib\site-packages\sklearn\metrics\_classification.py:1497: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
c:\users\hp\appdata\local\programs\python\python39\lib\site-packages\sklearn\metrics\_classification.py:1497: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


<h2>=> Độ chính xác F1 trung bình của mô hình DECTREE là 0.96</h2>

In [53]:
import pickle

# Save the trained model to a file using pickle
with open('dectree_HCV_model.pkl', 'wb') as file:
    pickle.dump(classifier, file)

Thử dự đoán 

In [54]:
# Age, ALB, ALP, ALT, AST, BIL, CHE, CHOL, CREA, GGT, PROT, isMale
X_new = [[50, 50, 50, 30, 25, 10, 8, 5, 80, 30, 70, True]]

classifier.predict(X_new)

array(['Hepatitis'], dtype=object)